### ISSUE TITLE
MP problems experienced when a remote user runs the old c172p model.

### ISSUE THREAD

**Issue by andgi** [Juanvvc/c172p-detailed#336]
Apparently (I don't use MP these days), there are some problems with the MP/AI model when a user with the new c172p(-detailed) receives data from a user that uses the old c172p model (e.g. using the current or a previous FG release).

It is possible to avoid some, but perhaps not all, of these problems by coding the conditions and expressions of the new animations defensively. If a required property is not sent its value will be nil, so if the nil value activates the normal non-damaged state and the default bushkit configuration the missing value would not be a problem. It's more difficult for MP properties that the new c172p uses for a different purpose than the old, but I don't know if there are many of these.

Since it is likely that there are quite a mix of FG versions on the MP network it is probably worth investing some time trying to remove some of the problems (e.g. that broken wings seems to be the default state).

An alternative and perhaps simpler solution is to attach some Nasal code to the c172p model that initializes all animation properties to a safe value. If the remote is a new c172p these values will be overwritten immediately, but if it is an old c172p the ones not sent will now have a safe value. It would still leave problems with MP properties that are used differently by both models, though.

**Comment by onox** [comment]
I don't think it's really unavoidable. It's a one time pain that you have to go through. 3.8 will probably be less painful.

To give an example, a lot of aircraft that use registration numbers (aka "immatriculation") use (waste) six generic ints (int[14] to int[19). What we did is use string[0] instead with about [six lines of Nasal code](https://github.com/Juanvvc/c172p-detailed/blob/master/Models/c172p.xml#L61-67) in `<nasal><load>`. Not only is our solution not wasting six integers, it's also more flexible; it's not necessarily limited to six digits.

**Comment by andgi** [comment]
Yes, some problems may be unavoidable but it is probably possible to avoid the the most visible ones, like broken wings. E.g. if the wing state is sent over a property/properties unused by the old c172p a few lines of Nasal in the model file could ensure that old c172p:s are not represented with broken wings.

**Comment by onox** [comment]
We could add a version number.

**Comment by andgi** [comment]
Yes, but wasting a MP property for that doesn't feel good. It might be enough to check the value of some property/ies that the new c172p sends and the old doesn't to separate them.

**Comment by wlbragg** [comment]
I'm not sure we should try to stay compatible with the original c172p. I thought the whole idea was that this replaces the original.

@onox, if we are going to try to maintain the original along side the new, you and I probably need to do am MP session, one old one new and log the problems. Then see if it is practical to fix.

**Comment by onox** [comment]
I think we can actually use string[1] for a version number. Most of the string[] properties are unused.

**Comment by andgi** [comment]
@onox Note that the amount of space in one MP packet is limited and that, AFAIK, there is no round-robin scheduling of which properties get sent if the space run out (there certainly wasn't any when I looked at the MP protocol 5-7 years ago). The encoding of string values in the MP protocol is also extremely wasteful (IIRC: 32 bits per character + a 32bit length field) so using one to send a constant value is something I would avoid.

**Comment by wlbragg** [comment]
I didn't know how to refer to a specific post in another issue so I will copy my thoughts here, FWIW.
>I am confused by why we are even discussing this farther, do we do this kind of backwards compatibility for anything in FG?
I guess this is a little different because it is the default aircraft, but not really. We shouldn't even attempt to mix the old with the new, as far as I am concerned.
The way I see it, if you want to fly the "old" c172p, you don't do it in MP and you don't do it in versions that it has been replaced by the "new" c172p (3.5+).
I came to that conclusion and opinion the second we were told to name the aircraft the same as the original, that is why at the time, I insisted on getting that question answered, because I knew it would make a difference in how we moved forward with this upgrade.
I can see a bit of the other side of this as 3.2-3.4 users still want to fly the c172p in MP, that is why i felt it important to give them the regression version that "should" be able to MP. Prior to 3.2, we need to let that go!
If 3.2-3.4 don't work correctly in MP, then I think we should look at that.

I feel you have to cut it off at some point, what was before the original c172p? I'll bet there was something before it, do we support whatever that was?

**Comment by wkitty42** [comment]
and mine from #305, too

>i think something is being misunderstood... the problem is that we, running fgfs 3.5, see older <=3.4 c172p craft with broken wings on our screens when they show up in our viewing area... i don't know what those older fgfses see when we show up... depending on our configuration, we may be floating if we have bush gear or pontoons on...
>
>i understand the reason for the regression version and i fully agree with the reasoning...
>
>i also understand not supporting older versions but this isn't about that... it is about how those old craft appear to us on our screens... we're not looking to support them... only to view them properly when they are in our view range...

i don't know if this can be done or not but it would be nice... granted, looking broken and being told they are looking like that may be incentive enough to get folks running old versions of fgfs to upgrade... especially if they want to do MP... only time will tell...

**Comment by andgi** [comment]
For FG it is highly unusual that an aircraft is completely replaced by another with the same name and even more so when it comes to the default aircraft. That makes the present situation special.

From my point of view, it is not a problem that a new/old or old/new combination over MP has less functionality than the old/old or new/new combinations. The problem is the highly visible regressions, like the broken wings, which will generate an endless stream of "bug" reports and complaints until FG 3.4 and older eventually (really eventually!) goes out of use.

**Comment by onox** [comment]
@andgi I'll see what I can do about those broken wings then.

**Comment by wlbragg** [comment]
What other base aircraft do we have and support with the same folder and set names?

@onox, if you need help over MP to evaluate all the potential problems, let me know and I will help.

**Comment by geoffmcl** [comment]
@onox, @andgi, @wlbragg this is the first positive news I have read ;=))

At this stage I think we are **ONLY** talking about release 3.4 c172p working with 3.5+ c172p over mp.

I will also help where I can. 

Being a maintainer of [fgms](https://gitlab.com/fgms/fgms-0-x), and the author of [crossfeed](https://gitlab.com/fgtools/crossfeed) I can, with a little setup, get access to the raw packet data as we run mpserver14.

But have read here and there others have done something similar with tcp/udp packet sniffers...

About a year back, I captured a giant raw packet log, with some 897,930 packets, and am working on an app to analyse those packets... WIP... Have already seen some interesting things, like prop IDS not in sIdPropertyList (48,49,...); packet average length was about 700 bytes back then; more to come...

I have three(3) machines, one of which, Dell01, is release 3.4 in a rather slow old XP machine. The others are Dell02 (linux) and Win7-pc (Windows 7) with the latest 3.5 build. 

And I think I also still have a 3.2 release build in Dell02...

Just tell me **where** and **when** you want to meet over mp, and I will add all or any of these to the mix... If I am here... Usually from about 11:00 to 22:00 CET (Paris) time...

Solving the c172p mp problems may also help understand the 707 engines on flame, but I think we should concentrate on the c172p, the **default** aircraft.

**Comment by andgi** [comment]
@geoffmcl How often do you see packets close to the maximum size - hinting that an overflow might have happened?
I was first a bit taken aback by your "author of crossfeed" statement until I saw that it was a separate tool from the feature in fgms. IIRC it was I that added the crossfeed option to fgms more than a few years ago. :)

**Comment by geoffmcl** [comment]
@andgi added some packet size stats, and it seems only a small percent approach the limit...

This is a crude quick table of packet sizes in my 897,930 raw packet log, circa Oct, 2012, abt 673 MB - 

```
Size 200 to 300 0.835032 percent...
Size 300 to 400 0.182754 percent...
Size 400 to 500 2.217879 percent...
Size 500 to 600 6.394931 percent...
Size 600 to 700 21.003085 percent...
Size 700 to 800 40.132527 percent...
Size 800 to 900 19.080329 percent...
Size 900 to 1000 5.299968 percent...
Size 1000 to 1100 4.853496 percent...
```

Given that none are in the max range 1100 to 1200 suggests that none were 'truncated', but that is not sure... as you can see in the FG code it precalculates the string size, and aborts if it exceeds MAX_PACKET_SIZE(1200)...

If you did indeed add the crossfeed feature to fgms, then thanks for a great 'feature' ;=))))

We use it to give us this json live list of pilots -

 http://crossfeed.freeflightsim.org/flights.json

And that feed is used to power my live mp map -

 http://geoffair.org/fg/map-test2/map-test.html

**Comment by andgi** [comment]
@geoffmcl Thanks. Good to know the packet sizes weren't that close to the limit. To continue off topic, back in the days I did add some features to fgms, you can see most of them among below the old_patches directory here http://www.gidenstam.org/FlightGear/fgms/ (LazyRelay, crossfeed and black list). I'm not sure to what extent these have been rewritten by now.

**Comment by wlbragg** [comment]
>@andgi I'll see what I can do about those broken wings then.

Is that the ONLY issue? It didn't sound like it according to what I read. Unless I got lucky, by chance, the gear and prop should be suspect to cause trouble also.

**Comment by andgi** [comment]
@wlbragg I'd say the broken wings is the no 1. issue. What I'm proposing is basically to try to cover the most visible compatibility problems as they are likely to give rise to many "bug" reports and complaints that somebody will need to respond to.

**Comment by onox** [comment]
I can turn the broken wings into chopped off wings by using int[17] instead of int[3]. Not sure yet how to get full wings.

**Comment by geoffmcl** [comment]
@wlbragg, @andgi I agree the BROKEN WINGS is the **main** issue here ;=()

Two more simple views -

http://geoffair.org/tmp/c172-3.4-view.png

http://geoffair.org/tmp/c172-3.5-view.png

The first is in the 'old' 3.4 c172, in 3.4 release looking at the remote c172 3.5 - no problem!

The second is in the 'new' 3.5 c172, in 3.5 latest looking at the remote c172 3.4 - broken wings. The prop and the gear looks fine...

Now to repeat, at first you see **nothing** in the 3.5 view!

To **make** the c172 model appear have to set menu View -> Adjust LOD Ranges -> AI/MP Aircraft to 0!

So this may not necessarily be an mp packet problem at all. It could be this setting of AI/MP to 0 does something else...

And just a time correction - my usual time at the computers is 11:00 to abt 20:00 CET (Paris) time, and will be closing tonight after this...

Leave anything here you want me to try tomorrow...

**Comment by onox** [comment]
This "AI/MP Aircraft to 0" is a bug in FG and totally unrelated to the c172p.

**Comment by andgi** [comment]
@onox I think it might be better if the new c172p MP model (via Nasal in the model file) discovered that the remote end is out-of-date and set some property that overrules the standard damage animations (e.g. their conditions could "and" their actual condition with not of the out-dated property - or make that a compatible flag instead..). With that approach you would not have to change so much of the intended damage animations in the new c172p.

**Comment by onox** [comment]
Yes, I will probably go for such a solution.

**Comment by wlbragg** [comment]
What about a simple NOT VOID. Isn't the 'new" c172p AI model looking for a value that is not there?

Also, what is the "old" c172p seeing over MP when flying with the "new"?

**Comment by onox** [comment]
I have added a possible fix to branch bug-280, so please test and report :smile:

**Comment by onox** [comment]
Fixed via PR #365.
